### PR TITLE
Apt get switch

### LINKED ISSUE #14: The "apt" Package provider should probably use "apt-get" instead of "apt"
apt is very noisy on output, and it would be nice to have something a bit more concise.

### PR DESCRIPTION
**Pull request by graysonhead**
This should fix #14 . I ended up needing to fix a few other things to properly test, but this switches out apt for apt-get with a flag that cleans up the output a bit.

You may want to look at 28f4765 specifically, as I repeated myself and I'm not sure the best way to un-repeat myself. My initial thought is to make an 'install_latest' signal that was called when version isn't specified, and only use 'install' when a version is specified. That would centralize the logic in Package as opposed to having to deal with that case in each of the methods, but then it would only make sense to create an 'upgrade_latest' as well. If you like the sound of that I think I can re-factor this PR although I won't be able to test brew.

### CODE REVIEW

**Comment by mpdehaan** [comment]
Hey thanks very much for this, this was quick and congrats on being the very first module PR!  If you want to possibly take a look at the second review comment that would be welcome, if not, I can make that happen Monday.  I'm not sure my first comment makes any sense, but that is clearly on me for not making that conditional very clear to start.  Appreciate the catch on that one.

While we aren't doing anything about updating the apt-cache, I would assume people know enough to just use the shell command to do it, and maybe we can add a note later about that.

(I don't have a good place for random notes in the module docs parser at the moment, so it would probably be something I'd add to the examples and comment out.  Also that last part is more of a note for myself too)

Thanks!!

**Comment by graysonhead** [comment]
> take a look at the second review comment that would be welcome, if not, I can make that happen Monday. I'm not sure my first comment makes any sense, but that is clearly on me for not making that conditional very clear to start. Appreciate the catch on that one.

I wouldn't mind taking a crack at the changes (free-time willing), but I can't seem to figure out where you left the review comments (I'm a bit new to GitHub's social features, so I'm probably just missing them.) I assume [this](https://help.github.com/articles/viewing-a-pull-request-review/) is how I would view them?

> While we aren't doing anything about updating the apt-cache, I would assume people know enough to just use the shell command to do it, and maybe we can add a note later about that.

I don't have a problem with the way everyone else does this (having an update_cache parameter that defaults to False.) Its not hard to implement it that way and results in a lot less copy paste when writing roles. I guess the right way to do that would be adding an 'update_cache' signal so that users can update the cache separately from installing/updating packages as well in case they have some other service/tool/Resource that interacts with packages and want to update the cache without modifying packages.

That would work for yum and apt anyways, I'm not sure about brew or some of the more niche Linux package managers.

**Comment by mpdehaan** [comment]
Sorry! that was my fault, GitHub makes you click "submit review" and I forgot to click... again :) I'll get used to this eventually.

update_cache is probably fine, yeah.  It could plan a seperate action, and you could still use update_cache standalone.

yum's cache is somewhat self updating, IIRC, so no immediate need to add it, but if we add an update_cache parameter to the type it could implement it later.

Thanks!  and yeah no problem if you don't have time, I can get to it later Monday most likely if you want, and just make those changes on top of what you have if needed - but if you want to just let me know, I mean what is there now *technically* works, so nobody is broken waiting on this.  Up to you!

File: opsmop/providers/package/package.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def plan(self):
         # FIXME: this can probably should advantage of the StrictVersion class to be smarter.
         if current_version and self.absent:
             self.needs('remove')
-        elif not current_version:
+        elif not current_version and not self.absent:
```
[the remark this answers is not in the document]
Comment (graysonhead) [reply opsmop/providers/package/package.py]: I think I see what you mean, I didn't realize it at the time but you can re-create this by doing:
`Package(name='lolcat', version='42.0.99-1', absent=True)`
That also will cause it to uninstall and reinstall on subsequent runs.

I think I understood what you were getting at with logic branches, and I think a27fc4d fixes this. If not feel free to correct it.

File: opsmop/providers/package/apt.py [hunk 2]
```diff
@@ -39,11 +39,17 @@ def apply(self):
         which = None
         if self.should('install'):
             self.do('install')
-            which = INSTALL.format(name=self.name)
+            if self.version:
+                which = INSTALL.format(name="{}={}".format(self.name, self.version))
```
[the remark this answers is not in the document]
Comment (graysonhead) [reply opsmop/providers/package/apt.py]: Sounds good. Implemented those suggestions in 41c816a

**Comment by graysonhead** [comment]
I just added two commits in response to your comments.

**Comment by graysonhead** [comment]
I took a stab at adding update_cache (at least, to apt for now), but ran into two issues:

It looks like apply is expected to return the results to the executor in order for the executor to evaluate whether or not the task failed, but there doesn't seem to be a provision for returning more than one task. The output looks like how I would expect it to:

```----------------------------------
1 :: AptTesting :: Package: lolcat

    planning...
        # dpkg -s lolcat | grep '^Version'
        | dpkg-query: package 'lolcat' is not installed and no information is available
        | Use dpkg --info (= dpkg-deb --info) to examine archive files,
        | and dpkg --contents (= dpkg-deb --contents) to list their contents.
        = ok, rc=1
    planned:
        | update_cache
        | install
    applying...
        # apt-get update -q=2
        | (no output)
        = ok, rc=0
        # apt-get -q=2 install -y lolcat=42.0.99-1
        | Selecting previously unselected package lolcat.
        | (Reading database ...
        | (Reading database ... 5%
        | (Reading database ... 10%
        | (Reading database ... 15%
        | (Reading database ... 20%
        | (Reading database ... 25%
        | (Reading database ... 30%
        | (Reading database ... 35%
        | (Reading database ... 40%
        | (Reading database ... 45%
        | (Reading database ... 50%
        | (Reading database ... 55%
        | (Reading database ... 60%
        | (Reading database ... 65%
        | (Reading database ... 70%
        | (Reading database ... 75%
        | (Reading database ... 80%
        | (Reading database ... 85%
        | (Reading database ... 90%
        | (Reading database ... 95%
        | (Reading database ... 100%
        | (Reading database ... 355625 files and directories currently installed.)
        | Preparing to unpack .../lolcat_42.0.99-1_all.deb ...
        | Unpacking lolcat (42.0.99-1) ...
        | Setting up lolcat (42.0.99-1) ...
        | Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
        = ok, rc=0
    actions:
        | update_cache
        | install
    ok, rc=0

complete!
```

However when I intentionally break the apt-upgrade command, the task still succeeds because the result for self.run(UPDATE_CACHE) doesn't get returned to the executor:

```----------------------------------
1 :: AptTesting :: Package: lolcat

    planning...
        # dpkg -s lolcat | grep '^Version'
        | dpkg-query: package 'lolcat' is not installed and no information is available
        | Use dpkg --info (= dpkg-deb --info) to examine archive files,
        | and dpkg --contents (= dpkg-deb --contents) to list their contents.
        = ok, rc=1
    planned:
        | update_cache
        | install
    applying...
        # apt-get update nb -q=2
        | E: The update command takes no arguments
        = fatal, rc=100
        # apt-get -q=2 install -y lolcat=42.0.99-1
        | Selecting previously unselected package lolcat.
        | (Reading database ...
        | (Reading database ... 5%
        | (Reading database ... 10%
        | (Reading database ... 15%
        | (Reading database ... 20%
        | (Reading database ... 25%
        | (Reading database ... 30%
        | (Reading database ... 35%
        | (Reading database ... 40%
        | (Reading database ... 45%
        | (Reading database ... 50%
        | (Reading database ... 55%
        | (Reading database ... 60%
        | (Reading database ... 65%
        | (Reading database ... 70%
        | (Reading database ... 75%
        | (Reading database ... 80%
        | (Reading database ... 85%
        | (Reading database ... 90%
        | (Reading database ... 95%
        | (Reading database ... 100%
        | (Reading database ... 355625 files and directories currently installed.)
        | Preparing to unpack .../lolcat_42.0.99-1_all.deb ...
        | Unpacking lolcat (42.0.99-1) ...
        | Setting up lolcat (42.0.99-1) ...
        | Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
        = ok, rc=0
    actions:
        | update_cache
        | install
    ok, rc=0

complete!
```

What is the best way to pass the results back to the executor when multiple actions are taken?

Secondly, calling `Package(update_cache=True)` fails as the Field class casts None from the name parameter into 'None' and then tries to install a package 'None':
```--------------------------------
1 :: AptTesting :: Package: None

    planning...
        # dpkg -s None | grep '^Version'
        | dpkg-query: package 'none' is not installed and no information is available
        | Use dpkg --info (= dpkg-deb --info) to examine archive files,
        | and dpkg --contents (= dpkg-deb --contents) to list their contents.
        = ok, rc=1
    planned:
        | update_cache
        | install
    applying...
        # apt-get update -q=2
        | (no output)
        = ok, rc=0
        # apt-get -q=2 install -y None
        | E: Unable to locate package None
        = fatal, rc=100
    actions:
        | update_cache
        | install
    fatal, rc=100
FAILED: fatal, rc=100
```

Using Package in this way (just to update the cache and nothing else) is probably pretty edge-case-ey, but I could see myself having used it once or twice back in the day. If we want this to be possible I suppose the two options are to either exempt "None" from type coercion in Field, or have package treat "None" as None (which seems potentially confusing).)

**Comment by mpdehaan** [comment]
Thanks for all of this, this looks good...

(A)


Comment on this part:

"It looks like apply is expected to return the results to the executor in order for the executor to evaluate whether or not the task failed, but there doesn't seem to be a provision for returning more than one task."

This isn't the case, thankfully.  If you use "self.run", a non-zero return code will result in an exception, so the self.apply() stage only needs to return self.ok(). The main reason modules like shell return the command result is that it can be used with "register" to save the output into a variable.  In the case of the apt module, the return value isn't interesting, so we don't really need to return anything other than self.ok(). Exception handling will deal with understanding that the command failed. I can take a look at making it understand that no plan actions need to happen when the package name is empty.

(B) the output  is one of the reasons (other than having access to more flags later) of using apt-get vs apt, is that apt was originally very noisy - so we probably need a way to silence all the Reading database lines. 

https://askubuntu.com/questions/258219/how-do-i-make-apt-get-install-less-noisy

I can probably fool around with this a bit, and worse case we can add some parameters to the command module about lines to suppress and not send to echo.

I'm fine with taking a look at both of those things.

**Comment by graysonhead** [comment]
> This isn't the case, thankfully. If you use "self.run", a non-zero return code will result in an exception, so the self.apply() stage only needs to return self.ok(). The main reason modules like shell return the command result is that it can be used with "register" to save the output into a variable. In the case of the apt module, the return value isn't interesting, so we don't really need to return anything other than self.ok()

In the case where I'm doing `Package(name='lolcat', version='42.0.99-1', update_cache=True)` and there are two planned tasks ('update_cache' and 'install'), should I be checking the return code for both steps within the self.apply() method and return self.fatal() if one of them failed? 

> the output is one of the reasons (other than having access to more flags later) of using apt-get vs apt, is that apt was originally very noisy - so we probably need a way to silence all the Reading database lines.

The Reading databases line are something that dpkg is doing, and since apt and apt-get both wrap that, they are going to show up either way unless you just silence all of stdout (or silence apt-get to the point where it doesn't give any useful output) https://bugs.launchpad.net/ubuntu/+source/apt/+bug/874409. Running with -q=2  on the terminal looks nicer as the Reading Database overwrites the previous line:

```
$ sudo apt-get -q=2 install lolcat
Selecting previously unselected package lolcat.
(Reading database ... 355625 files and directories currently installed.)
Preparing to unpack .../lolcat_42.0.99-1_all.deb ...
Unpacking lolcat (42.0.99-1) ...
Setting up lolcat (42.0.99-1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
```
What would be really neat, is if we could make OpsMop act more like a terminal and allow previous lines to be re-written, this would save module writers a ton of headache writing modules that call commands with "log unfriendly" output (like progress bars and such). I'm not sure how feasible that is though.


I won't be offended if you need to re-work all the update_cache stuff to your liking, I'm sure you have a vision in your head for the idiomatic way to do multi-step Resources that I lack (and I didn't look very hard, but there don't seem to be any examples in existing Resource modules.) I'm also starting to wonder if I should have submitted update_cache as a separate PR since it is only loosely related to the original issue I was trying to solve :).

**Comment by mpdehaan** [comment]
should I be checking the return code for both steps within the self.apply() method and return    
    self.fatal() if one of them failed?

No, you don't shouldn't need to do this.  Exception handling is automatically handled for any self.run() call that fails.  self.test() does not raise exceptions, but self.run() ultimately will. I think it could have been perhaps better if I rework *ALL* of the providers to never return anything, but that's not ideal because of the way self.register works, it's actually useful to register the result of the command class and that could also be true of other modules in the future.

    they are going to show up either way unless you just silence all of stdout (or silence apt-get to the 
    point where it doesn't give any useful output) 

Yep, that's what I was trying to say when I was linking to that article too - what I am proposing above is a feature where the self.run method will take a list of patterns, and if any output lines start with those patterns, it will ignore them.  I can take care of that part, it's pretty easy.

**Comment by mpdehaan** [comment]
I'm also starting to wonder if I should have submitted update_cache as a separate PR since it is only 
    loosely related to the original issue I was trying to solve with this PR :).

That part doesn't bother me, since with 'apt' there wasn't a need for update-cache, and switching to apt-get introduces the need for it, they are pretty related in my head anyway :)

**Comment by mpdehaan** [comment]
This is all merged in now, thanks!

I added an "ignore_lines" feature to the command class, which works well enough to hide the output from apt at least.

I also bumped the timeouts for allowing extremely longer package installs, which is something I missed earlier.

Since the returns from the command aren't meaningful, I just return self.ok all the time.

**Comment by mpdehaan** [comment]
Ha, you've caught something there, my fault for assuming it worked, we need to raise exceptions when those commands fail.

On it :)

**Pull request merged**